### PR TITLE
Add daenerys tool to perform maintenance tasks

### PR DESCRIPTION
**Pull request by austenmc**
(in progress)

### CODE REVIEW

**Comment by austenmc** [comment]
Ok @Vassyli, take a look at what I've got here.

File: src/Console/Command/ModuleRegisterCommand.php [hunk 1]
```diff
@@ -0,0 +1,46 @@
+<?php
+declare(strict_types=1);
+
+namespace LotGD\Core\Console\Command;
+
+use LotGD\Core\Bootstrap;
```
Comment (Vassyli) [inline src/Console/Command/ModuleRegisterCommand.php]: I would say we use first external libraries grouped by the library, then internal libraries.

File: src/EventManager.php [hunk 2]
```diff
@@ -63,21 +63,22 @@ public function publish(string $event, array &$context)
...
             $klass = new \ReflectionClass($class);
         } catch (\LogicException $e) {
             // Currently we do the same thing on not found as on some other
             // exception. Maybe we should do something different.
-            throw new ClassNotFoundException();
+            throw new ClassNotFoundException("Loading class ${class} failed");
         } catch (\ReflectionException $e) {
-            throw new ClassNotFoundException();
+            throw new ClassNotFoundException("Could not find class ${class}");
         }
```
Comment (Vassyli) [inline src/EventManager.php:82]: You still don't want to use class_exists?

File: src/Game.php [hunk 3]
```diff
@@ -20,6 +22,36 @@ public function __construct(
         $this->eventManager = $eventManager;
     }
 
+    public static function getVersion(): string
+    {
+        // Conforms to semantic versioning.
+        return '0.0.1';
```
Comment (Vassyli) [inline src/Game.php]: Shouldn't we use "master" or "dev" as the version? It's not really a released version yet.

Edit: If "dev" or "master" are not valid versions, we should follow the SemVer-Guide and use 0.1-dev instead.

File: bin/daenerys [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+    if (file_exists($file)) {
+        return include $file;
+    }
+}
+
+// Dance to find the autoloader.
+// TOOD: change this to open up the Composer config and use $c['config']['vendor-dir'] instead of "vendor"
+includeIfExists(getcwd() . '/vendor/autoload.php') ||
+includeIfExists(__DIR__ . '/../vendor/autoload.php') ||
+includeIfExists(__DIR__ . '/../autoload.php');
```
Comment (Vassyli) [inline bin/daenerys:14]: Why using such a strange construct? **DIR** should palways oint to the dir the file is in, even if called from directories above. All you need is the second line, right?

**Comment by Vassyli** [comment]
Looks good!

File: src/Game.php (hunk 3 above)
Comment (austenmc) [inline src/Game.php]: I can change it to 0.1.0-dev

File: src/EventManager.php (hunk 2 above)
Comment (austenmc) [inline src/EventManager.php:82]: I need to call `implementsInterface` so I need the `ReflectionClass` anyway.

File: bin/daenerys (hunk 4 above)
Comment (austenmc) [inline bin/daenerys:14]: No b/c if the package is installed from a local repo, it makes a symlink to the package, so `crate/bin/daenerys` is actually something like `path/to/core/bin/daenerys/` so `__DIR__ /../vendor/` is actually the vendor for the `path/to/core`.

File: src/Console/Command/ModuleRegisterCommand.php (hunk 1 above)
Comment (austenmc) [inline src/Console/Command/ModuleRegisterCommand.php]: Ah yes, will do. We should probably enforce sorting them too. Ill create an issue for making a linter that enforces this.

File: bin/daenerys (hunk 4 above)
Comment (Vassyli) [inline bin/daenerys:14]: ooooh, right. But the first line still is too much, isn't it?

Comment (austenmc) [inline bin/daenerys:14]: The most common use case is inside a crate, w/ the core installed as a dependency off packagist. In that case, you'd be running daenerys from `crate/`, and then `getcwd() . '/vendor/autoload.php'` is the same as `__DIR__ . '/../vendor/autoload.php'`.

But, if you installed the core off a local copy, then those directories are different, and the current working directory is the one you want, not the vendor inside of your `/path/to/core/`.

File: src/Game.php (hunk 3 above)
Comment (austenmc) [inline src/Game.php]: Actually, I think 0.1.0 is the right way: http://semver.org/

**Pull request merged**